site/en/blog/crx-scripting-api/index.md

```diff
@@ -0,0 +1,209 @@
+
+### Expanding scripting capabilities
+
+Another consideration that fed into the Manifest V3 design process was a desire to introduce
+additional scripting capabilities to Chrome's extension platform. Specifically, we wanted to add
+support for declarative content scripts and to expand the capabilities of the `executeScript`
+method.
+
+Declarative content scripts support has been an outstanding Chromium feature request for several
+years. Today, Manifest V2 and V3 Chrome extensions can only statically declare content scripts in
+their manifest.json file; the platform doesn't provide a way to register new content scripts, tweak
+content script registration, or unregister content scripts at runtime.
+
+While we knew that we wanted to tackle this feature request in Manifest V3, none of our existing
+APIs felt like the right home. We also considered aligning with Firefox on their [Content Scripts
+API][content-scripts], but very early on we identified a couple major drawbacks to this approach.
+First, we knew that we would have incompatible signatures (e.g. dropping`code` support). Second, our
+API had a different set of design constraints (e.g. needing a registration to persist beyond a
+service worker's lifetime). Finally, this namespace would also pigeonhole us to content script
+functionality where we're thinking about scripting in extensions more broadly.
+
+On the executeScript front, we also wanted to expand what this API could do beyond what the Tabs API version supported. More specifically, we wanted to support functions and arguments, more easily target specific frames, and target non-"tab" contexts.
+
+## Changes in Manifest V3
+
+In the previous section I mentioned that we're planning to add dynamic content script support. While
+I'd love to dig more into what this might look like or how it might work, design goals, or
+capabilities, unfortunately we're not quite ready to share those details.
+
+In the remainder of this post, I'd like to take a closer look at the similarities and differences between `chrome.tabs.executeScript` and `chrome.scripting.executeScript`.
+
+While considering how the platform would need to evolve in light of remotely hosted code
+restrictions, we wanted to find a balance between the raw power of arbitrary code execution and only
+allowing static content scripts. The solution we struck on was to allow extensions to inject a
+function as a content script and to pass an array of values as arguments.
+
+Let's take a quick look at an (oversimplified) example. Say we wanted to inject a script that
+greeted the user by name when the user clicks the extension's action button (icon in the toolbar).
+In Manifest V2, we could dynamically construct a code string and execute that script in the current
+page.
+
+```js
+// Manifest V2 extension
+let givenName = "Default";
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.tabs.executeScript({
+    code: `alert(\`Hello, ${givenName}\`)`
```

Comment:
```suggestion
    code: `alert('Hello, ${givenName}')`
```

Reply:
Revised this code block with a slightly different approach. The new sample fetches and evals a JS file, then extracts a property from the return value.